query-graphs-js/src/nonTrivialEdgePrecomputing.ts

```diff
@@ -0,0 +1,48 @@
+import { assert } from "@apollo/federation-internals";
+import { Edge, QueryGraph, QueryGraphState, simpleTraversal } from "./querygraph";
+
+export function preComputeNonTrivialFollowupEdges(graph: QueryGraph): (previousEdge: Edge) => readonly Edge[] {
+  const state = new QueryGraphState<undefined, readonly Edge[]>(graph);
+  simpleTraversal(graph, () => {}, (edge) => {
+    const followupEdges = graph.outEdges(edge.tail);
+    state.setEdgeState(edge, computeNonTrivialFollowups(edge, followupEdges));
+    return true;
+  });
+  return (previousEdge) => {
+    const nonTrivialFollowups = state.getEdgeState(previousEdge);
+    assert(nonTrivialFollowups, () => `Non-trivial followup edges of ${previousEdge} should have been computed`);
+    return nonTrivialFollowups;
+  }
+}
+
+function computeNonTrivialFollowups(edge: Edge, allFollowups: readonly Edge[]): readonly Edge[] {
+  switch (edge.transition.kind) {
+    case 'KeyResolution':
+      // After taking a key from subgraph A to B, there is no point of following that up with another key
+      // to subgraph C if that key has _the same_ conditions. This is because, due to the way key edges
+      // are created, if we have a key (with some conditions X) from B to C, then we are guaranteed to
+      // also have a key (with the same conditions X) from A to C, and so it's that later key we
+      // should be using in the first place. In other words, it's never better to do 2 hops rather than 1.
+      return allFollowups.filter((followup) => followup.transition.kind !== 'KeyResolution' || !sameConditions(edge, followup));
```

Comment:
> So if I'm reading this correctly (and going back to the diagram you showed us), if A->B uses key 'x', but A->C uses key 'y', a transition from B->C with key 'y' would not be filtered?

It's correct that it wouldn't be filtered by this method. But also, that case is fairly different because at the moment we add edges to the query graph without looking if their condition is "satisfiable" (it's the algorithm that checks it later) . That is, we may have a key 'y' between A->C but 'y' not being "reachable"/"collectable" from 'A'. Let's take an example:
```graphql
# subgraph A
type Query {
  t: T
}

type T {
  x: Int
}
```
```graphql
# subgraph B
type T @key(fields: "x") {
  x: Int
  y: Int
}
```
```graphql
# subgraph C
type T @key(fields: "y") {
  y: Int
  z: Int
}
```
In that example, the query graph will technically have the key edge "y" from A to C, but you cannot use directly because A doesn't not have "y". So to get "z", you genuinely have to got B then C. If anything, it's the A -> C edge that should be filtered.

And so the next question would be: why are adding edges like A->B for key "y" in the query graph is A doesn't have "y"? And we "might" (I haven't though about it deeply) be able to do something like that, but it's far from trivial because checking if a key "condition" can be satisfied can be pretty complex: it's trivial in the example above, but keys can have many field, including nested ones, and some of those field can well be external or have requires, etc..., so checking the satisfiability of a key condition essentially means running the validation algorithm on that condition. Again, it might be possible to:
1. create a query graph as we do today (so, with all key edges added regardless of the satisfiability of their condition).
2. run some "optimisation" step on that query graph that look at each key edge, see if it is ever satisfiable (essentially running a version of the validation algorithm on their condition), and remove them if they aren't.

We could them run the validation algorithm/query planning on that now "optimized" query graph. But I'll note that:
1. this would require quite a refactor of the current code. Definitively high effort.
2. I doubt it would speed up composition validation, because I don't think we would save work in that case, we'd just split the work in more steps, but it would sum up to the same work. That said ...
3. ... it would be likely beneficial to query planning, where the "optimized" query graph might be use for computing many query plans. How much it'll help in practice is unclear however.